### PR TITLE
App improvements

### PR DESCRIPTION
**Pull request by matteo-crypto**
## Explanation of Changes
- Highlighted the "search by text" feature.
- Various interface improvements.
- Finnish translation added.
- Ability to navigate back and forth for up to 20 courses.
- License added.
- Donate Us.

### CODE REVIEW

File: src/components/screens/main-screen.tsx [hunk 1]
```diff
@@ -86,6 +91,21 @@ const MainScreen: React.FC = () => {
     onCourseTextMatchDebounced(newQueryText);
   };
 
+  /**
+   * On course code change handler
+   * 
+  */
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:97]:
```suggestion
  /**
   * On course code change handler 
   */
```

File: src/components/screens/main-screen.tsx [hunk 2]
```diff
@@ -127,54 +147,129 @@ const MainScreen: React.FC = () => {
...
     return (
       <Stack
-        spacing={2}
-        component="form"
-        direction="row"
+        direction="column"
+        alignItems="center"
         width="100%"
+        spacing={4}
+        sx={{ padding: 2 }}
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:169]:
```suggestion
        spacing={4}
        padding={2}
```

File: src/components/screens/main-screen.tsx [hunk 3]
```diff
@@ -127,54 +147,129 @@ const MainScreen: React.FC = () => {
...
+   */
+  const renderExamples = () => (
+    <Stack direction="column" alignItems="center" spacing={2}>
+      <Typography variant="h4">{strings.mainScreen.examplesTitle}</Typography>
+      <Stack
+        direction="row"
+        spacing={2}
+        alignItems="stretch"
+      >
+        {strings.mainScreen.examples.map((example: string) => (
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:231]: Except for having these example strings in localization, put them at the top of the file as e.g.

QUERY_EXAMPLES=[
  "Example1",
  "Example2",
  "Example3",
]

Because we certainly don't want to have localization for example strings

File: src/components/screens/main-screen.tsx [hunk 4]
```diff
@@ -127,54 +147,129 @@ const MainScreen: React.FC = () => {
...
-      variant="outlined"
-      label={strings.mainScreen.courseQueryText}
-      onChange={({ target }) => onQueryTextChange(target.value)}
-    />
+    <Stack
+      direction="column"
+      alignItems="center"
+      width="100%"
+      spacing={4}
+      sx={{ padding: 2 }}
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:259]:
```suggestion
      spacing={4}
      padding={2}
```

File: src/components/screens/main-screen.tsx [hunk 5]
```diff
@@ -127,54 +147,129 @@ const MainScreen: React.FC = () => {
...
+      let commaIndex = option.name.indexOf(",");
+      commaIndex === -1 && (commaIndex = option.name.length);
+      let bracketIndex = option.name.indexOf("(");
+      bracketIndex === -1 && (bracketIndex = option.name.length);
+      const cutIndex = Math.min(commaIndex, bracketIndex);
+      option.name = option.name.substring(0, cutIndex);
+      if (option.name.length > 49) {
+        option.name = `${option.name.substring(0, 49)}...`;
+      }
+    });
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:161]: Why do we need to do this?

File: src/components/screens/main-screen.tsx [hunk 6]
```diff
@@ -199,7 +294,7 @@ const MainScreen: React.FC = () => {
    * Renders search
    */
   const renderSearch = () => (
-    <PaperCard elevation={6} sx={{ width: 650 }}>
+    <PaperCard elevation={6} sx={{ width: 700 }}>
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:297]: Why is this width changed

File: src/components/screens/main-screen.tsx [hunk 7]
```diff
@@ -231,11 +323,46 @@ const MainScreen: React.FC = () => {
   /**
    * Renders courses
    */
-  const renderCourses = () => (
-    <Stack spacing={3}>
-      {courses.map(course => <CourseCard course={course}/>)}
-    </Stack>
-  );
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:238]: Why is this removed?

File: src/components/screens/main-screen.tsx [hunk 8]
```diff
@@ -231,11 +323,46 @@ const MainScreen: React.FC = () => {
...
+     */
+    const handlePageChange = (_: React.ChangeEvent<unknown>, newPage: number) => {
+      setPage(newPage);
+      // Wait for a short delay before scrolling to the first course card on the new page
+      setTimeout(() => {
+        if (courseListRef.current !== null) {
+          courseListRef.current.scrollIntoView({ behavior: "smooth" });
+        }
+      }, 50);
+    };
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:342]: Put the handler before all the render methods

File: src/components/screens/main-screen.tsx [hunk 9]
```diff
@@ -231,11 +323,46 @@ const MainScreen: React.FC = () => {
...
+        <div ref={courseListRef}/>
+        {displayedCourses.map(course => (
+          <CourseCard key={course.code} course={course}/>
+        ))}
+        <Stack
+          direction="row"
+          sx={{
+            display: "flex",
+            justifyContent: "center"
+          }}
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:355]:
```suggestion
        <Stack
          direction="row"
          display="flex",
          justifyContent="center"
          
```

File: src/components/screens/main-screen.tsx [hunk 10]
```diff
@@ -231,11 +323,46 @@ const MainScreen: React.FC = () => {
...
+        >
+          <Pagination
+            count={Math.ceil(courses.length / coursesPerPage)}
+            page={page}
+            onChange={handlePageChange}
+          />
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:361]:
```suggestion
          { courses.length > coursesPerPage &&
            <Pagination
              count={Math.ceil(courses.length / coursesPerPage)}
              page={page}
              onChange={handlePageChange}
            />
          }
```

Show the pagination if the course count is more than one page (Check the lint, I am not 100% sure my code style complies with the linter)

File: src/components/screens/main-screen.tsx [hunk 11]
```diff
@@ -127,54 +147,129 @@ const MainScreen: React.FC = () => {
...
+        spacing={2}
+        alignItems="stretch"
+      >
+        {strings.mainScreen.examples.map((example: string) => (
+          <Button
+            key={example}
+            variant="outlined"
+            onClick={() => onQueryTextChange(example)}
+            sx={{
+              borderRadius: 1,
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:237]: Having border radius 1 is almost the same as 0. Either 0 or theme.spacing(1)

File: src/components/screens/main-screen.tsx [hunk 12]
```diff
@@ -127,54 +147,129 @@ const MainScreen: React.FC = () => {
...
+      <Typography variant="h4">{strings.mainScreen.examplesTitle}</Typography>
+      <Stack
+        direction="row"
+        spacing={2}
+        alignItems="stretch"
+      >
+        {strings.mainScreen.examples.map((example: string) => (
+          <Button
+            key={example}
+            variant="outlined"
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:234]:
```suggestion
            variant="text"
```

File: src/components/screens/main-screen.tsx (hunk 5 above)
Comment (matteo-crypto) [reply src/components/screens/main-screen.tsx:160]: Before, the items in the drop-down menu were almost all cut off because they all had too long text, now they are cropped to the right length

File: src/components/screens/main-screen.tsx (hunk 6 above)
Comment (matteo-crypto) [reply src/components/screens/main-screen.tsx:296]: To make it the same width as the courses. I thought it looked cleaner that way (visually). What do you think?

File: src/components/screens/main-screen.tsx (hunk 7 above)
Comment (matteo-crypto) [reply src/components/screens/main-screen.tsx:238]: Nono this chunck of code is present. I see your comment on the old version

File: src/components/screens/main-screen.tsx (hunk 5 above)
Comment (Chi-Chu319) [reply src/components/screens/main-screen.tsx:180]: I'll add comments

File: src/components/screens/main-screen.tsx (hunk 6 above)
Comment (Chi-Chu319) [reply src/components/screens/main-screen.tsx:311]: I wanted to have different widths to make the search box visually distinguished. But since you have changed it, I'll leave it

File: src/components/screens/main-screen.tsx [hunk 13]
```diff
@@ -83,9 +93,38 @@ const MainScreen: React.FC = () => {
...
+      setSearchMode(SearchMode.TEXT);
+      setCourseCode("");
+    }
+    setCourses([]);
+    setShowExamples(true);
+  };
+
+  /**
+   * Handles page change
+   */
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:117]:
```suggestion
  /**
   * Handles page change
   * 
   * @param newPage new page index
   */
```

File: src/components/screens/main-screen.tsx [hunk 14]
```diff
@@ -127,54 +166,124 @@ const MainScreen: React.FC = () => {
...
+              }}
+            >
+              {limitation}
+            </Button>
+          ))}
+        </Stack>
+      </AccordionDetails>
+    </Accordion>
+  );
+
```
Comment (Chi-Chu319) [inline src/components/screens/main-screen.tsx:238]:
```suggestion
```

File: src/localization/en.json [hunk 15]
```diff
@@ -44,8 +44,16 @@
...
+    "courseCodeDescription": "Pick a course from the list to get the most matched courses.",
+    "courseQueryTextPlaceholder": "Send a message...",
+    "matchYourCourses": "Match your courses!",
+    "examplesTitle": "Ask anything or try an example",
+    "limitationsTitle": "Limitations",
+    "limitations": [
+      "May occasionally generate incorrect information",
+      "Only courses taught in English are included in the search",
+      "Accepts only English text for the search"
+    ]
```
Comment (Chi-Chu319) [inline src/localization/en.json:57]:
```suggestion
```

File: src/localization/fi.json [hunk 16]
```diff
@@ -40,12 +40,20 @@
...
+    "courseCodeDescription": "Valitse kurssi luettelosta saadaksesi sopivimmat kurssit.",
+    "courseQueryTextPlaceholder": "Lähetä viesti...",
+    "matchYourCourses": "Sovita kurssit yhteen!",
+    "examplesTitle": "Kysy mitä tahansa tai kokeile esimerkkiä",
+    "limitationsTitle": "Rajoitukset",
+    "limitations": [
+      "Saattaa toisinaan tuottaa virheellisiä tietoja",
+      "Vain englanninkieliset kurssit sisältyvät hakuun",
+      "Hyväksyy vain englanninkielisen tekstin hakuun"
+    ]
```
Comment (Chi-Chu319) [inline src/localization/fi.json:57]:
```suggestion
```

File: src/localization/strings.ts [hunk 17]
```diff
@@ -67,9 +66,13 @@ export interface Localized extends LocalizedStringsMethods {
...
     moreOptions: string;
-    courseCode: string;
-    courseQueryText: string;
+    courseCodePlaceholder: string;
+    courseCodeDescription: string;
+    courseQueryTextPlaceholder: string;
     matchYourCourses: string;
+    examplesTitle: string;
+    limitationsTitle: string;
+    limitations: string[];
```
Comment (Chi-Chu319) [inline src/localization/strings.ts:75]:
```suggestion
```
